File: expense-tracker/routes/tags.py
from flask import Blueprint, render_template, request, redirect, url_for, session, flash, jsonify
from database.db import get_db

tags_bp = Blueprint('tags', __name__)


@tags_bp.route('/tags', endpoint='manage_tags')
def manage_tags():
    if 'user_id' not in session:
        return redirect(url_for('auth.login'))
    db = get_db()
    user_id = session['user_id']
    tags = db.execute("SELECT t.*, COUNT(et.expense_id) as usage_count FROM tags t LEFT JOIN expense_tags et ON t.id = et.tag_id WHERE t.user_id = ? GROUP BY t.id ORDER BY usage_count DESC, t.name", (user_id,)).fetchall()
    return render_template('tags.html', tags=tags)


@tags_bp.route('/tags/add', methods=['POST'], endpoint='add_tag')
def add_tag():
    if 'user_id' not in session:
        return redirect(url_for('auth.login'))
    name = (request.form.get('name') or '').strip()
    color = request.form.get('color', '#6366f1')
    if not name:
        flash('Tag name is required.', 'danger')
        return redirect(url_for('tags.manage_tags'))
    db = get_db()
    try:
        db.execute("INSERT INTO tags (user_id, name, color) VALUES (?, ?, ?)", (session['user_id'], name, color))
        db.commit()
        flash(f'Tag "{name}" created!', 'success')
    except db.IntegrityError:
        flash(f'Tag "{name}" already exists.', 'danger')
    return redirect(url_for('tags.manage_tags'))


@tags_bp.route('/tags/<int:id>/delete', methods=['POST'], endpoint='delete_tag')
def delete_tag(id):
    if 'user_id' not in session:
        return redirect(url_for('auth.login'))
    db = get_db()
    db.execute("DELETE FROM expense_tags WHERE tag_id = ?", (id,))
    db.execute("DELETE FROM tags WHERE id = ? AND user_id = ?", (id, session['user_id']))
    db.commit()
    flash('Tag deleted.', 'info')
    return redirect(url_for('tags.manage_tags'))


@tags_bp.route('/api/tags', endpoint='api_tags')
def api_tags():
    if 'user_id' not in session:
        return jsonify([])
    db = get_db()
    tags = db.execute("SELECT id, name, color FROM tags WHERE user_id = ? ORDER BY name", (session['user_id'],)).fetchall()
    return jsonify([dict(t) for t in tags])


@tags_bp.route('/tags/expense/<int:expense_id>/set', methods=['POST'], endpoint='set_expense_tags')
def set_expense_tags(expense_id):
    if 'user_id' not in session:
        return redirect(url_for('auth.login'))
    db = get_db()
    expense = db.execute("SELECT * FROM expenses WHERE id = ? AND user_id = ?", (expense_id, session['user_id'])).fetchone()
    if not expense:
        flash('Expense not found.', 'danger')
        return redirect(url_for('dashboard.dashboard'))
    tag_ids = request.form.getlist('tag_ids')
    db.execute("DELETE FROM expense_tags WHERE expense_id = ?", (expense_id,))
    for tid in tag_ids:
        try:
            db.execute("INSERT INTO expense_tags (expense_id, tag_id) VALUES (?, ?)", (expense_id, int(tid)))
        except:
            pass
    db.commit()
    flash('Tags updated!', 'success')
    return redirect(request.referrer or url_for('dashboard.dashboard'))

File: backend/routes/tags.js
const express = require('express');
const router = express.Router();
const mongoose = require('mongoose');
const { Tag, Expense } = require('../models');
const { is_valid_hex_color } = require('../src/helpers');

router.get('/tags', async (req, res, next) => {
    if (!req.session.user_id) {
        return res.redirect('/login');
    }
    const user_id = req.session.user_id;

    try {
        const tags = await Tag.find({ user_id }).lean();

        const usageCounts = await Expense.aggregate([
            { $match: { user_id: new mongoose.Types.ObjectId(user_id) } },
            { $unwind: '$tags' },
            { $group: { _id: '$tags', count: { $sum: 1 } } }
        ]);
        const usageMap = {};
        for (const uc of usageCounts) {
            usageMap[uc._id.toString()] = uc.count;
        }

        const tags_data = tags.map(tag => ({
            ...tag,
            id: tag._id.toString(),
            usage_count: usageMap[tag._id.toString()] || 0
        }));

        // Sort by usage_count DESC, then name ASC
        tags_data.sort((a, b) => {
            if (b.usage_count !== a.usage_count) {
                return b.usage_count - a.usage_count;
            }
            return a.name.localeCompare(b.name);
        });

        res.render('tags.html', { tags: tags_data });
    } catch (err) {
        next(err);
    }
});

router.post('/tags/add', async (req, res, next) => {
    if (!req.session.user_id) {
        return res.redirect('/login');
    }
    const user_id = req.session.user_id;
    const name = (req.body.name || '').trim();
    const color = req.body.color || '#6366f1';

    if (!name) {
        req.flash('danger', 'Tag name is required.');
        return res.redirect('/tags');
    }

    if (!is_valid_hex_color(color)) {
        req.flash('danger', 'Invalid tag color.');
        return res.redirect('/tags');
    }

    try {
        const newTag = new Tag({
            user_id,
            name,
            color
        });
        await newTag.save();
        req.flash('success', `Tag "${name}" created!`);
    } catch (err) {
        if (err.code === 11000) {
            req.flash('danger', `Tag "${name}" already exists.`);
        } else {
            return next(err);
        }
    }
    res.redirect('/tags');
});

router.post('/tags/:id/delete', async (req, res, next) => {
    if (!req.session.user_id) {
        return res.redirect('/login');
    }
    const user_id = req.session.user_id;
    const tag_id = req.params.id;

    if (!mongoose.isValidObjectId(tag_id)) {
        req.flash('danger', 'Tag not found.');
        return res.redirect('/tags');
    }

    try {
        // Pull tag from all user expenses
        await Expense.updateMany(
            { user_id: new mongoose.Types.ObjectId(user_id) },
            { $pull: { tags: new mongoose.Types.ObjectId(tag_id) } }
        );

        // Delete tag
        await Tag.deleteOne({ _id: tag_id, user_id });
        req.flash('info', 'Tag deleted.');
    } catch (err) {
        return next(err);
    }
    res.redirect('/tags');
});

router.get('/api/tags', async (req, res) => {
    if (!req.session.user_id) {
        return res.json([]);
    }
    try {
        const tags = await Tag.find({ user_id: req.session.user_id }).sort({ name: 1 }).lean();
        res.json(tags.map(t => ({
            id: t._id.toString(),
            name: t.name,
            color: t.color
        })));
    } catch (err) {
        res.status(500).json({ error: err.message });
    }
});

router.post('/tags/expense/:expense_id/set', async (req, res, next) => {
    if (!req.session.user_id) {
        return res.redirect('/login');
    }
    const user_id = req.session.user_id;
    const expense_id = req.params.expense_id;

    if (!mongoose.isValidObjectId(expense_id)) {
        req.flash('danger', 'Expense not found.');
        return res.redirect('/dashboard');
    }

    let tag_ids = req.body.tag_ids || [];
    if (!Array.isArray(tag_ids)) {
        tag_ids = [tag_ids];
    }
    const validTagIds = tag_ids.filter(t => mongoose.isValidObjectId(t));

    try {
        const expense = await Expense.findOne({ _id: expense_id, user_id });
        if (!expense) {
            req.flash('danger', 'Expense not found.');
            return res.redirect('/dashboard');
        }

        expense.tags = validTagIds;
        await expense.save();

        req.flash('success', 'Tags updated!');
        res.redirect(req.get('referrer') || '/dashboard');
    } catch (err) {
        next(err);
    }
});

module.exports = router;
